src/cpu/o3/dyn_inst.hh

```diff
@@ -158,6 +158,8 @@ class DynInst : public ExecContext, public RefCounted
         SquashedInIQ,            /// Instruction is squashed in the IQ
         SquashedInLSQ,           /// Instruction is squashed in the LSQ
         SquashedInROB,           /// Instruction is squashed in the ROB
+        CannotExecute,           /// Processor does not have capability to
```

Comment:
It is probably not important but I can't figure out if this should reside in the Status enum or in the Flags enum.... Looking at the enum cases it looks like the initial idea was to track the progress/status of the instruction in the pipeline. However later cases are using this to carry instruction behaviour (looking at the Serialize options for example) regardless of the position in the pipeline.

I believe in the future we should clean this up. Anyway I am not going to push strongly on this as it needs refactoring anyway. One reason to prefer Flags is that the CannotExecute is sort of misleading if checked against CanIssue description... I would honestly consider removing any reference to the exectution bit as the instruction will actually execute in gem5 (it will call StaticInst::execute), it will only abort if the cpu tries to commit it on a later stage. So maybe using NoCapableFU (similarly to what we are already using in the IQ)?

Reply:
Agree with you on the Status vs. Flags. 

However I do not follow you with the `StaticInst::execute`. With the patch, if O3CPU does not have a capable function unit, the instruction does not issue and so `inst->execute();` in iew.cc will not be called. If you prefer I change it that is fine, but unless I'm mistaken in my understanding of how issue and iew work in O3CPU the instruction cannot and will not execute (with the path).

Reply:
> However I do not follow you with the `StaticInst::execute`. With the patch, if O3CPU does not have a capable functional unit, the instruction does not issue and so `inst->execute();` in iew.cc will not be called.

Yes you are right I was somehow convinced you were retaining old behaviour (instruction would issue), only you were panicking when trying to commit the instruction.

I would still try to issue for the reason mentioned above:

> The problem is that most instructions produce the undefined instruction fault only once they get executed. At the head of the execute logic they would check for the feature presence and return a fault if feature is not supported/disabled. If we prevent those instructions from being issued, we will never produce a fault and simulation will be stuck..

Now what happens if you run a simulation with a CPU without SIMD FUs and therefore SIMD disabled in the ISA?

* Real HW: The instruction won't be decoded and undefined instruction will be triggered
* Issue even if no FU: matching the HW; we issue/execute even with no FU and if the ISA logic is correct we trigger undefined instruction. If the ISA does not generate an exception, we abort at commit as a safety net.
* Not issueing (your implementation): You just abort and do not give a chance to the instruction to generate a fault during execute. So this forces the user to provision the O3CPU with SIMD FUs even if they are not using it...

So I would treat NoCapableFU as NoNeedFU, only I would flag the instruction to panic if we try to commit it.
Please let me know your thoughts

Reply:
I know it is a niche case, but sometimes arch correctness means correctly handling wrong behaviour ;)

Reply:
I see your point, but then I can just call inst->execute() functionally when it fails to find a capable FU. I assume on real HW it would not take up an issue slot (maybe not even an IQ entry) ?
 
Something like (inst_queue.cc):

```
 } else if (idx == FUPool::NoFreeFU) {
            iqStats.statFuBusy[op_class]++;
            iqStats.fuBusy[tid]++;
            ++order_it;
        } else {
            assert(idx == FUPool::NoCapableFU);
            // CPU has no capable FU for the instruction
            // but this may be OK if the instruction gets
            // squashed. 
            // First, execute the instruction to give a chance to
            //  the CPU to behave as real hardware would and
            // trigger an illegal instruction fault that
            // will be carried to commit.
            // If no fault is triggered, remember there is no FU
            // and act if the instruction is identified as being part of
            // the correct path (reaches head of ROB)
            // In both cases, keep it in the IQ:
            // Fault will do an inclusive squash and
            // panic won't care
            
            issuing_inst->execute();

            // ISA did not trigger the fault
            // Remember it, commit will panic()
            if (issuing_inst->getFault() == NoFault)
            {
                issuing_inst->setCannotExecute();
                issuing_inst->setCanCommit();
             }

              // Do not try re-issuing it again
              listOrder.erase(order_it++);
        }
```

Reply:
I just don't particularly like the idea of fast-tracking the faulting instruction. This method is supposed to simply schedule the ready instructions; adding this specific case will complicate the pipeline model IMHO. Also, even assuming the instruction takes an issue slot, it does not seem seem important to me as either simulation will panic or future instructions will be squashed anyway (by the exception).

Reply:
> By "fast tracking", do you mean the case where the instruction is at the head of the ROB and the fault would trigger one cycle earlier than if the instruction had issued and gone to IEW to do the call to `execute()`? Because it will not fault until it reaches head of commit regardless of when it issues here (just so we are on the same page).

No by fast-tracking I don't mean I am concerned about the early execution;  I am concerned about the software engineering aspect of providing two different routes that go from point A (issue) to point B (execute).

Basically I don't like the extra addition of the execute( call in the instruction queue)

```
} else {
  assert(idx == FUPool::NoCapableFU);
  // CPU has no capable FU for the instruction
  // but this may be OK if the instruction gets
  // squashed. 
  // First, execute the instruction to give a chance to
  // the CPU to behave as real hardware would and
  // trigger an illegal instruction fault that
  // will be carried to commit.
  // If no fault is triggered, remember there is no FU
  // and act if the instruction is identified as being part of
  // the correct path (reaches head of ROB)
  // In both cases, keep it in the IQ:
  // Fault will do an inclusive squash and
  // panic won't care
            
  issuing_inst->execute();

  // ISA did not trigger the fault
  // Remember it, commit will panic()
  if (issuing_inst->getFault() == NoFault)
  {
    issuing_inst->setCannotExecute();
  }
```

It basically as if we are forking the pipeline at the IQ point (from a SW engineering POV) and let the faulting instruction flow in this "fast-track". I think the faulting instruction should follow the existin flow.

> 
> By letting the instruction be issued rather than treated here, the code here will just have to be split between inst_queue.cc (remember it did not get an FU, even if it issues) and iew.cc (execute(), and if fault, nothing, if no fault, tell commit to panic when instruction reaches commit).
> 

Sorry I don't understand what modification we need to make to iew.cc
What I thought would suffice is to treat NoCapableFU like NoFreeFU (they would reuse the same code as they have been doing until now), only in the first case we would annotate the instruction to panic at commit. 
Let me know if there's something I am missing

Reply:
> Sorry I don't understand what modification we need to make to iew.cc
What I thought would suffice is to treat NoCapableFU like NoFreeFU (they would reuse the same code as they have been doing until now), only in the first case we would annotate the instruction to panic at commit.
Let me know if there's something I am missing

No I think you're right, I was thinking I would also need to change the iew stage to mark that the instruction was marked by IQ and did not fault, but commit can do it. I'll go with this, thanks for the feedback.

Reply:
Alright, so I have changed the flag name in dyn_inst.hh. That flag is set in inst_queue.cc if there is no capable FU but the instruction will still issue. At commit, the instruction will either have faulted, which will go through the usual fault handling code, or will not, in which case commit will panic. 

The underlying assumption is that if a fault is not triggered, the instruction will still do enough things to be marked as commitable, but I think that's a reasonable assumption.